stub: Add support for .ucode EFI addons

tfg13 opened the pull request:
Hello!

This is the promised followup for #31872, and .ucode section support to addon files.

I implemented largely as discussed with @poettering here: https://github.com/uapi-group/specifications/pull/98

We allow one `.ucode` section per UKI, and addon file and then merge them in the following order:

1. ~~UKI embedded .ucode section~~
2. ~~Global addons~~
3. ~~UKI addons~~

~~Note this order is different from what was discussed in the thread above. The reason for that is that both cmdline and devicetree do it in this order, so I figured this is more consistent. (And makes sense IMHO, if one drops an addon file to change a (distro-)shipped UKI).~~

~~This is now also aligned with the measurement order.~~

[Edit]: This changed again, see below for context. The order is now UKI addons, Global addons, UKI embedded.

























































<!-- devel-freezer = {"comment-id":"2088279696","freezing-tag":"v256-rc1"} -->

tfg13 commented:
I still need to do the manpages and docs, but looking for input on the questions above first

tfg13 commented:
I changed the order now and updated the initial comment to reflect that. I think the order makes more sense this way, is the same as with cmdline and devicetree addons, and also matches the measurement order. Let me know what you think.

YHNdnzj commented:
> I changed the order now and updated the initial comment to reflect that. I think the order makes more sense this way, is the same as with cmdline and devicetree addons, and also matches the measurement order. Let me know what you think.

Hmm, so I guess the original idea is that since kernel stops on first match, and addons are the way to override/extend what's set in the UKI, we still need the ucodes in addons to have higher priorities.

I'm not familiar with ucode, but for cmdline, the same option defined later overrides the former ones. So in that case appending what's from addon to the cmdline from UKI makes sense, but not so much for ucodes.

YHNdnzj commented:
cc @poettering

tfg13 commented:
Ok, I reworked it again. The order for microcode updates is now:

1. UKI addons
2. Global addons
3. UKI embedded section

They are also measured in this order.

The measurement part is a bit tricky and we might want to talk about it. Two things:
- Since the embedded section is now last, I am skipping it in the initial measurement loop that otherwise measures all the UKI sections. It gets measured later in the code. (This techincally violates the comment in uki.h a bit, that says the order defined there is always the measurement order. I don't have a good alternative here, I think we must measure in the order the files will be used in the end, no?)
- As is its right now, the measurement of UKIs with embedded .ucode sections is not correctly predictable for systemd-pcrlock. From what I can tell, this is also true for other addons like cmdline, (I think pcrlock doesn't full handle them, right?), but it might be worse here, since .ucode sections are fundamentally initrds, so they go into the initrd register.

poettering commented on src/boot/efi/stub.c:
```diff
@@ -735,6 +884,24 @@ static EFI_STATUS run(EFI_HANDLE image) {
+        /* Handle ucode. Always call this, even if there are no addons.
+         * This measures the .ucode section embedded in the UKI (if any). */
+        err = ucode_merge_and_measure_addons(
+                ucode_bases_addons_global,
```
as per coding style, we indent line-broken function calls to twice the usual indentation, i.e. 16 spaces

poettering commented on src/fundamental/uki.h:
```diff
@@ -24,6 +24,8 @@ extern const char* const unified_sections[_UNIFIED_SECTION_MAX + 1];
+         * the measurement, and hence shouldn't be input to it.
+         * The .ucode section is not included because it is measured by systemd-stub separately later,
+         * after merging with addon files. */
+        return section >= 0 && section < _UNIFIED_SECTION_MAX && section != UNIFIED_SECTION_PCRSIG && section != UNIFIED_SECTION_UCODE;
```
i don't think there's a problem with measuring this even if this is not used in the end because it is overriden.

poettering commented on src/boot/efi/stub.c:
```diff
@@ -341,6 +341,138 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+
+        if (n_ucode_addons_global == 0 && n_ucode_addons_uki == 0 && ucode_section_size == 0) {
+                return EFI_SUCCESS;
+        }
```
coding style: no {} around single-line if blocks

poettering commented on src/boot/efi/stub.c:
```diff
@@ -341,6 +341,138 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+        /* Measure UKI addons */
+        for (size_t i = 0; i < n_ucode_addons_uki; i++) {
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_bases_addon_uki[i]), ucode_sizes_addon_uki[i], UNIFIED_SECTION_UCODE, &sections_measured);
+        }
```
similar

poettering commented on src/boot/efi/stub.c:
```diff
@@ -341,6 +341,138 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+        /* Measure global addons */
+        for (size_t i = 0; i < n_ucode_addons_global; i++) {
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_bases_addon_global[i]), ucode_sizes_addon_global[i], UNIFIED_SECTION_UCODE, &sections_measured);
+        }
```
ditto

poettering commented on src/boot/efi/stub.c:
```diff
@@ -341,6 +341,138 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+        /* Measure .ucode section from UKI */
+        if (ucode_section_size > 0) {
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_section_base), ucode_section_size, UNIFIED_SECTION_UCODE, &sections_measured);
+        }
```
ditto

poettering commented on src/boot/efi/stub.c:
```diff
@@ -341,6 +341,138 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+
+        /* Merge the combined global addons, combined UKI addons, and ucode section from UKI */
+        err = combine_initrds(
+                (const void*const[]) {
```
indent to 16ch

poettering commented:
> The measurement part is a bit tricky and we might want to talk about it. Two things:
> 
>     * Since the embedded section is now last, I am skipping it in the initial measurement loop that otherwise measures all the UKI sections. It gets measured later in the code. (This techincally violates the comment in uki.h a bit, that says the order defined there is always the measurement order. I don't have a good alternative here, I think we must measure in the order the files will be used in the end, no?)

So as mentioned in my review comments, double measurement is not really a problem, in particular for shorter stuff. it's better to measure too much then too little.

>     * As is its right now, the measurement of UKIs with embedded .ucode sections is not correctly predictable for systemd-pcrlock. From what I can tell, this is also true for other addons like cmdline, (I think pcrlock doesn't full handle them, right?), but it might be worse here, since .ucode sections are fundamentally initrds, so they go into the initrd register.

Yes, this is currently not covered by pcrlock, we should add that one day. But for now this shouldn't stop this PR.

poettering commented:
Could you rebase on top of #33461 right-away? it contains a rework of the "measured" flag handling, which this work of yours also touches

tfg13 commented:
Thanks for the review! Will fix the style, put the double-measurement back (which will simplify things), and rebase.

tfg13 commented:
Rebased again after #33461 was merged

tfg13 commented on src/boot/efi/stub.c:
```diff
@@ -360,6 +360,137 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+                global_addons_combined_base = PHYSICAL_ADDRESS_TO_POINTER(global_addons_combined_pages.addr);
+        }
+
+        /* Measure .ucode section from UKI again */
+        if (ucode_section_size > 0)
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_section_base), ucode_section_size, UNIFIED_SECTION_UCODE, &sections_measured);
```
This was important before when I excluded the UKI embedded `.ucode` section from the initial measurement. But now it is measured there - does measuring it again here give us anything? Ucode is tiny, it shouldn't really cost anything, opinions on keep vs kick out?

poettering commented:
I fear my recent rework of stub.c requires a major rebase of this PR. Sorry!

tfg13 commented:
No worries, I'm rewriting it at the moment

poettering commented on src/boot/efi/stub.c:
```diff
@@ -298,6 +328,71 @@ static void cmdline_append_and_measure_addons(
+
+        if (*n_ucode_addons == 0) {
+                return EFI_SUCCESS;
+        }
```
no {} around single line if blocks

tfg13 replied:
I really need a linter /o\

poettering commented on src/boot/efi/stub.c:
```diff
@@ -298,6 +328,71 @@ static void cmdline_append_and_measure_addons(
+        if (err != EFI_SUCCESS)
+                return err;
+
+        // copy the combined ucode initrd back into the initrds array to leave things in a consistent state for later cleanup
```
coding style: please use `/* … */` style comments for explanatory stuff. We use `// …` style comments for fixme style stuff only.

poettering commented on src/boot/efi/stub.c:
```diff
@@ -298,6 +328,71 @@ static void cmdline_append_and_measure_addons(
+        /* Measure just the ucode addons, the uki-embedded one was already measured. */
+        FOREACH_ARRAY(addon, *ucode_addons, *n_ucode_addons) {
+                measure_section(POINTER_TO_PHYSICAL_ADDRESS(addon->iov_base), addon->iov_len, UNIFIED_SECTION_UCODE, sections_measured);
+        }
```
remove {}, as above

poettering commented on src/boot/efi/stub.c:
```diff
@@ -298,6 +328,71 @@ static void cmdline_append_and_measure_addons(
+                iovec_done(i);
+
+        free(a);
+}
```
i think it would make sense to add this to src/fundamental/iovec-util-fundamental.h as generic inline func, maybe called iovec_free_many() or so?

poettering commented on src/boot/efi/stub.c:
```diff
@@ -298,6 +328,71 @@ static void cmdline_append_and_measure_addons(
+                struct iovec *tmp = ucode_addons[i];
+                ucode_addons[i]= ucode_addons[*n_ucode_addons - 1 - i];
+                ucode_addons[*n_ucode_addons - 1 - i] = tmp;
+        }
```
maybe split this out into a generic helper iovec_array_reverse() or so?

poettering commented on src/boot/efi/stub.c:
```diff
@@ -912,6 +1017,11 @@ static EFI_STATUS run(EFI_HANDLE image) {
+        /* Combine ucode initrds with the embedded initrd */
+        err = ucode_merge_and_measure_addons(initrds, &ucode_addons, &n_ucode_addons, &sections_measured);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error merging ucode initrds: %m");
```
I'd really prefer if we'd only combine initrds once instead of first combining the ucode initrds, and then combining it with the rest of the initrds.

Hence, change ucode_merge_and_measure_addons() so that it gets the fixed-size initrds[] array as input, and that it then returns a new array of initrd iovecs, which contains the ucode initrds as prefix, and then the fixed-sizes initrds[] copied into the back.

or in other words: instead of combining the actual initrd *payload* here you'd just combine the iovec array.

tfg13 replied:
aaaah right that makes sense. I tried to find an elegant solution but didn't get very far. Let me try

poettering commented on src/boot/efi/stub.c:
```diff
@@ -474,6 +572,16 @@ static EFI_STATUS load_addons(
                         };
                 }
+
+                if (ucode_addons && PE_SECTION_VECTOR_IS_SET(sections + UNIFIED_SECTION_UCODE)) {
```
mm, i think we should probably apply the uname conditioning above only to cmdline+dtb, but not to the ucode part, because ucode really shouldn't have an efefct on kernel compatibility.

tfg13 replied:
hmm sure, I can change that - but it makes things a bit more complex and inconsistent if some addons behave differently?

tfg13 replied:
Maybe: Ucode updates are sometimes rolled out for CPU mitigations (spectre/meltdown and friends). Those often go hand in hand with kernel mitigations, maybe that is a reason why someone would want to tie these two together?

poettering commented on src/boot/efi/stub.c:
```diff
@@ -91,6 +91,36 @@ static void combine_measured_flag(int *value, int measured) {
+                        unified_sections[section],
+                        &m);
+        combine_measured_flag(sections_measured, m);
+}
```
so please don't measure the ucode stuff with this. Please use tagged event measurement instead, see how this is done for dt. i.e. make up a new tag like DEVICETREE_ADDON_EVENT_TAG_ID does it, and then measure it with that.

the measurement of sections we do so far we must retain for compat, but it's actually very much not fun to dissect later from the measurement logs, hence we shouldn't add more to that.

poettering commented:
btw, would it be too much to ask to add support for .initrd add-ons at the same time? it would be pretty much the identical code, except for the ordering of things (i.e. ucode cpios should be placed in reverse order before UKI initrds, and initrd cpios should be placed in regular order after UKI initrds), but it's otherwise pretty much the same logic, no?

There's an old PR for this in #28070, but work on it stalled. I am pretty sure the folks who want that would be very thankful if you add this logic to your PR and fix it for them.

(but it#s fine if you don't make this your problem, but i thought I'd ask)

tfg13 commented:
Reworked and addressed all comments. It looks much better now, and handling a dynamic list of iovecs even has the nice bonus that `initrds_need_combine` can be replaced with a simple if.

tfg13 commented on src/boot/efi/stub.c:
```diff
@@ -912,16 +976,18 @@ static EFI_STATUS run(EFI_HANDLE image) {
         lookup_embedded_initrds(loaded_image, sections, initrds);
 
+        /* Measures ucode addons and populates all_initrds with the final order of all initrds */
+        measure_ucode_addons_and_create_final_list(initrds, &all_initrds, &n_all_initrds, ucode_addons, n_ucode_addons, &sections_measured);
```
Is `sections_measured` the right one to use here?

poettering replied:
no. you are measuring into TPM2_PCR_KERNEL_CONFIG, hence us parameters_measured.

tfg13 commented:
Re: Adding multi-initrd support. It should be super easy now, happy to try - but not in this PR, this is already going on for waaay too long.

tfg13 commented:
Rebased on the mem fixes

DaanDeMeyer commented on src/boot/efi/stub.c:
```diff
@@ -357,14 +356,69 @@ static void install_addon_devicetrees(
         }
 }
 
+static inline void iovec_array_append(struct iovec **arr, size_t *n_arr, struct iovec elem) {
```
Add `assert(arr)` and `assert(n_arr)` here.

DaanDeMeyer replied:
Maybe name this `iovec_array_extend()` since we have `strv_extend()` as well to do the same for strv.

DaanDeMeyer commented on src/boot/efi/stub.c:
```diff
@@ -357,14 +356,69 @@ static void install_addon_devicetrees(
+                NamedAddon *ucode_addons,
+                size_t n_ucode_addons,
+                int *sections_measured) {
+        EFI_STATUS err;
```
We always add a space after a multiline function declaration before we begin variable declarations

poettering commented on src/boot/efi/stub.c:
```diff
@@ -357,14 +356,69 @@ static void install_addon_devicetrees(
+
+        /* Now append the other initrds from the static list */
+        FOREACH_ARRAY(i, initrds, _INITRD_MAX) {
+                if (iovec_is_set(i))
```
maybe move this check into iovec_array_append(), so that you do an early exit there if the iovec is no set.

poettering commented on src/boot/efi/stub.c:
```diff
@@ -863,8 +922,12 @@ static EFI_STATUS run(EFI_HANDLE image) {
+        NamedAddon *dt_addons = NULL;
         size_t n_dt_addons = 0;
+        NamedAddon *ucode_addons = NULL;
+        size_t n_ucode_addons = 0;
```
erge these two lines with the preceeding two lines?

poettering commented on src/boot/efi/stub.c:
```diff
@@ -863,8 +922,12 @@ static EFI_STATUS run(EFI_HANDLE image) {
         size_t n_dt_addons = 0;
+        NamedAddon *ucode_addons = NULL;
+        size_t n_ucode_addons = 0;
+        struct iovec *all_initrds = NULL;
```
`_cleanup_free`

poettering commented on src/boot/efi/stub.c:
```diff
@@ -357,14 +356,69 @@ static void install_addon_devicetrees(
+        FOREACH_ARRAY(i, initrds, _INITRD_MAX) {
+                if (iovec_is_set(i))
+                        iovec_array_append(all_initrds, n_all_initrds, *i);
+        }
```
i'd split this loop into a separate function that you call from the main func.

poettering commented on src/boot/efi/stub.c:
```diff
@@ -912,16 +976,18 @@ static EFI_STATUS run(EFI_HANDLE image) {
-                /* If we have generated initrds dynamically or there is a microcode initrd, combine them with
-                 * the built-in initrd. */
-                err = combine_initrds(initrds, _INITRD_MAX, &initrd_pages, &final_initrd.iov_len);
+        if (n_all_initrds > 1) {
```
i think it would make sense to change the line `final_initrd = initrds[INITRD_BASE]` line in the `else` branch that belongs to this `if` to `final_initrd = all_initrds[0]` here. Would of course not actually make any difference, but I think it makes things clearer.

poettering commented on src/fundamental/tpm2-pcr.h:
```diff
@@ -49,3 +49,6 @@ enum {
 #define DEVICETREE_ADDON_EVENT_TAG_ID UINT32_C(0x6c46f751)
+
+/* The tag used for EV_EVENT_TAG event log records covering ucode addons (effectively initrds) */
+#define UCODE_ADDON_EVENT_TAG_ID UINT32_C(0xdac08e1a)
```
Please document this briefly in docs/TPM2_PCR_MEASUREMENTS.md, just copy the entry for the devicetree stuff, and make minor adjustments

poettering commented:
looks really good. I think we are close

DaanDeMeyer commented on src/boot/efi/stub.c:
```diff
@@ -912,16 +976,18 @@ static EFI_STATUS run(EFI_HANDLE image) {
-                /* If we have generated initrds dynamically or there is a microcode initrd, combine them with
-                 * the built-in initrd. */
-                err = combine_initrds(initrds, _INITRD_MAX, &initrd_pages, &final_initrd.iov_len);
+        if (n_all_initrds > 1) {
```
Is this actually true with https://github.com/uapi-group/specifications/pull/83?

tfg13 replied:
Hmm good point. Now that I think about it, was the stub already broken for UKIs without any initrds?

Before this PR, we would pass an empty iovec, and with this PR we pass NULL to `linux_exec` - both of them will fail the same on `assert(iovec_is_valid(initrd));` in `linux_exec`.

I am happy to pass something else if that is useful, but I don't want to also put fixing `linux_exec` in scope here, as is is kinda unrelated. Is that fair?

DaanDeMeyer replied:
That's fair, let's keep it as is then.

poettering commented:
lgtm

bluca commented:
`systemd-boot: Assertion 'ucode_addons' failed at ../src/boot/efi/stub.c:381@measure_and_append_ucode_addons, halting.`

bluca commented on src/boot/efi/stub.c:
```diff
@@ -357,14 +356,79 @@ static void install_addon_devicetrees(
+
+        assert(all_initrds);
+        assert(n_all_initrds);
+        assert(ucode_addons);
```
`assert(ucode_addons || n_ucode_addons == 0)`

tfg13 replied:
d'oh

YHNdnzj commented on src/boot/efi/stub.c:
```diff
@@ -463,15 +527,28 @@ static EFI_STATUS load_addons(
+                if (ucode_addons && PE_SECTION_VECTOR_IS_SET(sections + UNIFIED_SECTION_UCODE)) {
+                        *ucode_addons = xrealloc(*ucode_addons,
+                                                 *n_ucode_addons * sizeof(NamedAddon),
+                                                 (*n_ucode_addons + 1)  * sizeof(NamedAddon));
```
doubled space?

DaanDeMeyer replied:
You can drop the double spaces, it's just a typo from a previous PR
